MAINTAINERS.md

```diff
@@ -1,6 +1,108 @@
+    - named like "$prefix:[_**release short label**_](#assign-release-parameters)" per [agoric-3-proposals: Naming](https://github.com/Agoric/agoric-3-proposals/pkgs/container/agoric-3-proposals#naming) (conventionally using "a" for the unreleased $prefix, e.g. `a:upgrade-8`)
+    - containing a **package.json** having an object-valued `agoricProposal` property with `sdkImageTag` set to "unreleased" and `planName` set to the [_**release label**_](#assign-release-parameters)
+    - containing other files appropriate for the upgrade per [agoric-3-proposals: Files](https://github.com/Agoric/agoric-3-proposals/pkgs/container/agoric-3-proposals#files)
+
+    For example, see the [upgrade-14 PR](https://github.com/Agoric/agoric-sdk/pull/8755).
+
+### Populate the "dev release" branch
+
+For each set of changes to include:
+- [ ] Create a work branch from the "dev release" branch.
+- [ ] Cherry-pick approved changes from `master` onto this work branch.
+  - Avoid any cherry-pick that has side-effects on code you do not wish to be upgraded. For example, a change impacting deployed vat code that is not meant to be part of the upgrade (comment/types only changes may be acceptable, but will still result in new bundles, so should be avoided if possible). Chain software code is usually safe to upgrade as long as integration with code running inside vats doesn't change. For example, the host side of any bridge and the set of swingset syscalls must remain backwards compatible.
+  - For integrating a single PR (best suited to vat code changes which often require manual adaptation):
+    - [ ] Cherry-pick the commits of the `master` PR onto the work branch with minimal/mechanical conflict resolutions.  
+      It's acceptable to author the changes on the "dev release" based work branch and then rebase them to a master PR, as long as the PRs lands at "about the same time" (`master` PR **MUST** be merged before a release is cut).
+    - [ ] As needed, author any extra commits required for compatibility on the "dev release" work branch.
+    - [ ] Create a PR with the "dev release" branch as target.
+    - [ ] Reviewer must check that changes are narrowly focused, and do not affect unrelated components.
+      - For vat code PRs, changes to other packages tests, tools, or scripts may be acceptable.
+      - Compare the resulting code of the changed component with its code on master, and highlight any unexpected differences.
+  - For integrating a set of PRs (best suited to PRs that do not affect vat code and can be cherry-picked somewhat cleanly):
+    - [ ] Optional: Ensure [git rerere](https://git-scm.com/book/en/v2/Git-Tools-Rerere) ("reuse recorded resolution") is locally enabled:
+      ```sh
+      git config rerere.enabled true
+      ```
+    - [ ] Author a `rebase-todo` by selecting relevant changes from the `master` branch rebase log (see below).
```

Comment:
I still owe the regex to build the todo log

Reply:
Here is how I've been generating the todo log:

- Switch to `master` branch
- `git rebase -i --rebase-merges $LAST_COMMIT_IN_PREVIOUS_LOG`
- Run the following substitution to "organize" the todo log
  - `\n# Branch (.*)\nreset .*\n(merge -C .*\n)?((?:pick .*\(#\d+\)\n)*)(?:label branch-point.*\n)?((?:pick .*\n)*)(?:merge -C .*\n)?(?:label \1\n)` -> `$2\n$3\n# Branch $1\nlabel base-$1\n$4label $1\nreset base-$1\n`
  - Manually fix the last block
  - Add comments for squashed PRs
- Run the rebase and verify the commit graph does not change (aka that the re-organization didn't change anything)

Reply:
Was this useful for dev-upgrade-15?

Reply:
Hmm, I did not recognize your todo lists as derived from something produced by git and instead just manually based #9264 and #9266 on what you had done with upgrade-14 PRs like #8942. But even looking now, I don't think I understand where `$LAST_COMMIT_IN_PREVIOUS_LOG` is coming from, or how those "reorganizations" would be applied. Is the latter scriptable in a way that could be explained and checked in?

Reply:
> Is the latter scriptable in a way that could be explained and checked in?

A possibility: restructure blocks, then rename labels.

<details>

```sh
awk '
  # Restructure branch-specific blocks from `git rebase -i --rebase-merges`
  # so that `merge` commands appear at block end rather than block start.

  NR == 1 && match($0, /^label /) {
    # An initial `label` (but not a following no-op `reset`) belongs in the first block.
    firstBlockPrefix = $0 "\n";
    noopReset = "reset " substr($0, RLENGTH + 1, length($0) - RLENGTH);
    next;
  }
  firstBlockPrefix != "" && $0 == noopReset {
    next;
  }
  /^$|^# Branch .*/ {
    # TODO: We may be able to find the PR number from a commit id or branch name.
    sub(/Branch/, "PR #??? Branch");
    blockHeader = blockHeader $0 "\n";
    next;
  }
  !match($0, /^(reset|merge -C) /) {
    printf "%s%s", blockHeader, firstBlockPrefix;
    blockHeader = "";
    firstBlockPrefix = "";
  }
  {
    print;
  }
  END {
    printf "%s%s", blockHeader, firstBlockPrefix;
  }
' | awk '
  # Rename each label that receives a `merge` to "base-$branchName".

  function addLabel(label) {
    if (++labels[label] == 1) return;
    print "duplicate label: " label > "/dev/stderr";
    exit 1;
  }
  match($0, /^$|^# .*[Bb]ranch /) {
    branch = substr($0, RLENGTH + 1, length($0) - RLENGTH);
  }
  branch != "" && match($0, /^merge -C /) && match(prev, /^reset /) {
    onto = substr(prev, RLENGTH + 1, length($0) - RLENGTH);
    sub(/ .*/, "", onto);
    newOnto = "base-" branch;
    addLabel(newOnto);
    renames[++renameCount] = onto SUBSEP newOnto;
  }
  /^label / {
    addLabel(substr($0, RLENGTH + 1, length($0) - RLENGTH));
  }
  {
    buf = buf $0 "\n";
    prev = $0;
  }
  END {
    # For "last-wins" semantics, apply renames in reverse order.
    for (i = renameCount; split(renames[i], pair, SUBSEP); i--) {
      for (j = split("label reset", rebaseCmds, " "); cmd = rebaseCmds[j]; j--) {
        newBuf = "";
        seekLine = sprintf("\n%s %s", cmd, pair[1]);
        while (k = index(buf, seekLine)) {
          len = length(seekLine);
          r = sprintf("\n%s %s", cmd, pair[2]);
          if (!match(substr(buf, k + len, 1), /[ \n]/)) r = seekLine;
          newBuf = newBuf substr(buf, 1, k - 1) r;
          buf = substr(buf, k + len, length(buf) - (k - 1) - len);
        }
        buf = newBuf buf;
      }
    }
    printf "%s", buf;
  }
'
```

</details>